### PR TITLE
Basic restructuring and configuration options

### PR DESCRIPTION
**Pull request by davebshow**

### CODE REVIEW

**Comment by davebshow** [comment]
This is about ready to merge. Feel like doing a quick review @leifurhauks? The file loading should probably be a bit more robust...

**Comment by davebshow** [comment]
Ah, this branch probably needs the timeout config too, I forgot. Willfix.

**Comment by davebshow** [comment]
I think this is ready to merged to master...comments or suggestions @leifurhauks or @jsenecal?

**Comment by davebshow** [comment]
Doh! Forgot to update the session.py module....

File: goblin/driver/cluster.py [hunk 1]
```diff
@@ -0,0 +1,200 @@
...
+    A cluster of Gremlin Server hosts. This object provides the main high
+    level interface used by the :py:mod:`goblin.driver` module.
+
+    :param asyncio.BaseEventLoop loop:
+    """
+
+    DEFAULT_CONFIG = {
+        'scheme': 'ws',
+        'hosts': ['localhost'],
+        'port': 8182,
```
Comment (leifurhauks) [inline goblin/driver/cluster.py:48]: I think it should be possible for different hosts to use different ports. We could either have the hosts be in the form 'host:port' or make each host a tuple or dictionary like `('localhost', 8182)` or `{'url': 'localhost', 'port': 8182}`.

As an example, if the app connects to several hosts through the same firewall or load balancer, you could have multiple hosts with the same hostname but different ports.

**Comment by leifurhauks** [comment]
What's happening with `Goblin.traversal_source`/`session.traversal_source`? It seems that it's currently being passed to the server with the key `traversal_source`, not `aliases`:

``` json
{
  "requestId": "dfc266bb-bac0-4c8f-965d-10f3c6ca844e",
  "op": "eval",
  "args": {
    "gremlin": "graph.features().graph().supportsTransactions()",
    "traversal_source": {
      "g": "GRAPH_SERVICE.g"
    }
  },
  "processor": ""
}
```

**Comment by leifurhauks** [comment]
So it looks like there are a lot of places, especially in `Session`, where you pass a kwarg `traversal_source` to submit; but I think you dropped that kwarg from `Connection.submit` in 652a08f. As a result, all of those methods are generating requests that are missing the aliases.

Also, it looks like you started implementing `Client` so that it could keep track of the alias(es), which would be helpful so that it doesn't have to be passed to submit every time; but so far `Client` isn't doing anything with its `_traversal_source` attribute.

Also, on DSE, there is no default graph/traversal source on the server side. So almost any request will fail if it doesn't have a custom alias set. Currently `Goblin.open` doesn't take a kwarg for aliases or traversal_source, and it calls the `supports_transactions()` method which immediately fails because `graph` doesn't exist unless you set a custom alias.

What I propose is that connections, obtained with `Cluster.get_connection`, should only ever be used by instances of `Client`. That goes for `supports_transactions`, and also on `Session` objects -- `Session` objects should have a `Client` rather than a conn. That was the `Client` can keep track of the alias(es), which is simpler than having to keep track of them in every object that uses a connection, and having to pass them to `submit` in so many different places.

The `Client` could also keep track of other settings that get sent with requests, like `processor`, `mime_type` and so on. In other words, if I don't pass processor to `client_instance.submit(...)`, the processor used is the one set on that client. Passing it explicitly to submit would override that. I think that would be useful because those parameters are likely to be reused with every request (or almost).

File: goblin/driver/cluster.py (hunk 1 above)
Comment (leifurhauks) [reply goblin/driver/cluster.py:48]: I just realized gremlin-console's `remote.yaml` works this way too -- all hosts must use the same port. Seems odd to me, but I guess it's alright.

**Comment by davebshow** [comment]
I think bddc4e3 fixes all of the aliases issues you raised in your comment. Note that the structure of the `submit` function has been redesigned to accept a wider variety of arguments. This will make even more sense when I merge the GraphSON2 branch, which is currently WIP.

**Comment by leifurhauks** [comment]
I'll install this branch on Monday and try it out, but based on looking over the diffs, the new changes look good.

**Comment by davebshow** [comment]
Hey thanks @leifurhauks for taking some Saturday time on this. I'm going to add support for new gremlin-python as well as GraphSON2 this weekend, so I'll branch from here and merge back to the `config` branch. We can review on Monday, merge to `master`, and release. Question: how do you want to handle the new processor/op tests? Should we update the docker image to use TinkerPop 3.2.2? No rush, we can determine on Monday if you want.

**Pull request merged**